master/internal/core_experiment.go

```diff
@@ -484,7 +482,7 @@ func (m *Master) parseCreateExperiment(params *CreateExperimentParams, user *mod
 
 	project, err := getCreateExperimentsProject(m, params, user, defaulted)
 	if err != nil {
-		return nil, nil, false, nil, err
+		return nil, config, nil, false, nil, err
```

Comment:
I dislike pointless pointers, because it conveys incorrect type information.

Something that returns a `*Thing, err` might actually return `nil, nil`.  It _probably_ doesn't.  But why allow for the nil case at all?

Fwiw, I followed this all over with the experiment config.  So if you want pointers everywhere... let me know soon before I make it worse haha

Reply:
i dislike non-pointers because it conveys incorrect type information

something that returns `Thing, err` appears to always return a valid `Thing`. it _probably_ doesn't. but why allow for the user to get confused and try to use it at all. in code i write, (*T, err) kind of implies `Result<T>`.  also returning T makes it more normal to return partial failures which i hate. 

lol, so this is a joke in format, but is how i feel. though, tbh, not super strongly. i think there is a perf aspect here that favors you, and it really adds up if you do this everywhere. so as long as we don't do partial failures i'm cool with it. i also don't follow my own advice for primitive values, just slices/maps/most structs.

Reply:
I think `Result<T>` is a great way to discuss this.  Also I think this is an interesting discussion topic, so I'll explain my position at length.

Across multiple languages, I operate under the implicit assumption that, except in extremely rare cases, when there's an error returned, all other values are in an undefined state.  In some languages that assumption is enforced (like `Result` in rust or exception control flows in python) but even where it is not enforced, it's by far the most common idiom.  And under that assumption, I see returning `Thing, err` as identical to a `Result<Thing>`, while returning `*Thing, err` is actually an annoying `Result<*Thing>`.

Let's imagine writing C code that is guaranteed-memory-safe, for the `Thing, err` equivalent case:

```c
// returns nonzero for error
int dostuff(thing_t *thing);

int main(){
  thing_t t;
  int ret = dostuff(&t);
  if(ret){
    // error case
    return ret;
  }
  // no more checks; t is good to go
  return thing_method(&t);
}
```

and for the `*Thing, err` case:
```c
// returns nonzero for error
int dostuff(thing_t **thing);

int main(){
  thing_t *t;
  int ret = dostuff(&t);
  if(ret){
    // error case
    return ret;
  }
  if(!t){
    // wtf is this case for?
    fprintf(stderr, "dostuff didn't fail but also didn't give a valid t\n");
    abort();
  }
  // now t is good to go
  return thing_method(t);
}
```

Similar situation in the python case, to make mypy happy.  Here for `Thing, err` equivalent:


```python
# raises ValueError sometimes 
def dostuff() -> Thing:
  ...

thing = dostuff()
# thing is ready to go
thing.method()
```

but for the `*Thing, error` case, you need an extra assert:

```python
# raises ValueError sometimes
def dostuff() -> Optional[Thing]:
  ...

thing = dostuff()
# thing might be None... but why?
assert thing is not None, "internal logic error, thing is None"
# now you can use thing
thing.method()
```

I'm not fluent in rust, but I understand in rust you'd use switch statements basically identically.